components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -101,179 +101,110 @@ base::flat_map<std::string, std::string> MakeBraveServicesKeyHeader() {
-    json_rpc_service_->GetSolanaTokenAccountsByOwner(
-        account_address,
-        base::BindOnce(&AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner,
-                       weak_ptr_factory_.GetWeakPtr(),
-                       std::move(barrier_callback)));
-  }
-}
-
-void AssetDiscoveryManager::OnGetSolanaTokenAccountsByOwner(
-    base::OnceCallback<void(std::vector<SolanaAddress>)> barrier_callback,
-    const std::vector<SolanaAccountInfo>& token_accounts,
-    mojom::SolanaProviderError error,
-    const std::string& error_message) {
-  if (error != mojom::SolanaProviderError::kSuccess || token_accounts.empty()) {
-    std::move(barrier_callback).Run(std::vector<SolanaAddress>());
-    return;
-  }
-
-  // Add each token account to the all_discovered_contract_addresses list
-  std::vector<SolanaAddress> discovered_mint_addresses;
-  for (const auto& token_account : token_accounts) {
-    // Decode Base64
-    const absl::optional<std::vector<uint8_t>> data =
-        base::Base64Decode(token_account.data);
-    if (data.has_value()) {
-      // Decode the address
-      const absl::optional<SolanaAddress> mint_address =
-          DecodeMintAddress(data.value());
-      if (mint_address.has_value()) {
-        // Add the contract address to the list
-        discovered_mint_addresses.push_back(mint_address.value());
-      }
-    }
-  }
-
-  std::move(barrier_callback).Run(discovered_mint_addresses);
-}
-
-void AssetDiscoveryManager::MergeDiscoveredSolanaAssets(
-    bool triggered_by_accounts_added,
-    const std::vector<std::vector<SolanaAddress>>&
-        all_discovered_contract_addresses) {
-  // Create vector of all discovered contract addresses
-  std::vector<std::string> discovered_mint_addresses;
-  for (const auto& discovered_contract_address_list :
-       all_discovered_contract_addresses) {
-    for (const auto& discovered_contract_address :
-         discovered_contract_address_list) {
-      discovered_mint_addresses.push_back(
-          discovered_contract_address.ToBase58());
-    }
-  }
-
-  // Convert vector to flat_set
-  base::flat_set<std::string> discovered_mint_addresses_set(
-      std::move(discovered_mint_addresses));
-
-  auto internal_callback = base::BindOnce(
-      &AssetDiscoveryManager::OnGetSolanaTokenRegistry,
-      weak_ptr_factory_.GetWeakPtr(), triggered_by_accounts_added,
-      std::move(discovered_mint_addresses_set));
-
-  // Fetch SOL registry tokens
-  BlockchainRegistry::GetInstance()->GetAllTokens(mojom::kSolanaMainnet,
-                                                  mojom::CoinType::SOL,
-                                                  std::move(internal_callback));
+      base::BarrierCallback<std::vector<mojom::BlockchainTokenPtr>>(
+          3,
+          base::BindOnce(&AssetDiscoveryTask::MergeDiscoveredAssets,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+  DiscoverSPLTokensFromRegistry(sol_account_addresses,
```

Comment:
Need to pass sol_chain_ids too if we expect to support multi for Solana, otherwise at least make it clear it's mainnet only atm. Maybe leave a comment here saying this one is mainnet only atm.